### PR TITLE
Added service annotation to set Azure DNS label for public IP

### LINKED ISSUE #44775: [Azure] Support setting the DNS name label for public IPs created by ingress controllers
**What keywords did you search in Kubernetes issues before filing this one?** ingress controller dns name

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): FEATURE REQUEST

When deploying an ingress controller on Azure, Kubernetes helpfully creates a public IP address.

It would be useful to set the "DNS name label" of the created IP address based on an annotation in the Deployment. CNAMEs can then be used to point to this Azure DNS name, resulting in a (more) stable DNS configuration if the ingress controller is deleted and recreated.

**Kubernetes version** (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.4", GitCommit:"7243c69eb523aa4377bce883e7c0dd76b84709a1", GitTreeState:"clean", BuildDate:"2017-03-07T23:53:09Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.3", GitCommit:"029c3a408176b55c30846f0faedf56aae5992e9b", GitTreeState:"clean", BuildDate:"2017-02-15T06:34:56Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: Azure
- **OS** (e.g. from /etc/os-release): Ubuntu 16.04.2
- **Kernel** (e.g. `uname -a`): Linux k8s-master-30143214-0 4.4.0-66-generic #87-Ubuntu SMP Fri Mar 3 15:29:05 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
- **Install tools**: -
- **Others**: -

Downstream issue https://github.com/Azure/acs-engine/issues/517

### PR DESCRIPTION
**Pull request by tomerf**
**What this PR does / why we need it**: Added a feature to set the DNS label for public IPs in the Azure cloud.
For example:
```
apiVersion: v1
kind: Service
metadata:
  annotations:
    service.alpha.kubernetes.io/label-name: myservice
...
```
Will resolve myservice.westus.cloudapp.azure.com to the service's IP.

**Which issue this PR fixes**: fixes #44775

**Special notes for your reviewer**: Note that this is defining a new annotation, so feel free to point out if there is a preferred convention or anything else that needs to be done.

**Release note**:
```release-note
New service annotation "service.beta.kubernetes.io/azure-dns-label-name" to set Azure DNS label name for public IP
```

### CODE REVIEW

**Comment by tomerf** [comment]
Any comments on the new feature?

**Comment by quinton-hoole** [comment]
@tomerf Is there any reason why this should apply only to Azure?
Also, is there a design document for this?  It seems like the provision of DNS records for Ingress could/should be generalized, and not done one-off for a single provider.
I'd like @kubernetes/sig-network-feature-requests to take a look at this, and approve a design before merging the relevant PR.

**Comment by tomerf** [comment]
I'm working with Azure and could really use this functionality. I am not familiar with how other clouds implement this. I can research for other vendors if the network team agrees. I would also welcome suggestions/contributions from someone with the relevant experience.

Using annotations seems to me as the most reasonable approach at this level, and could use another review if generalized.

**Comment by quinton-hoole** [comment]
@tomerf Yes, I imagine that if you proposed a design for exposing DNS for ingress in general, that would be well received.  I do recall that there were some challenges with doing that (which is why it was not done initially).  I don't remember the details though. @kubernetes/sig-network-feature-requests might be able to help.  Or try their slack channel.
https://github.com/kubernetes/community/wiki/SIG-Network

**Comment by tomerf** [comment]
Ok, I have written a basic proposal. I listed a few use cases and some more ideas (like a static IP).
Can you provide feedback and let's continue from there?
[Design proposal](https://gist.github.com/tomerf/be23e33432596c3c40188e3d91535859)

**Comment by cmluciano** [comment]
@tomerf Can you please submit this proposal as a PR against https://github.com/kubernetes/community/blob/master/contributors/design-proposals/ . You can then add the ```/sig network``` label so that SIG-Network maintainers can review it.

**Comment by tomerf** [comment]
@quinton-hoole I submitted a design proposal (kubernetes/community#983) but the sig-network team doesn't think this feature is generic enough, so I think we should pursue this idea for Azure only.
Maybe in the future someone will come with a more comprehensive idea (support for third party DNS services?).

File: pkg/api/annotation_key_constants.go
Comment (thockin) [inline pkg/api/annotation_key_constants.go]: Cloud-specific labels should be defined in cloud-specific places

Comment (tomerf) [reply pkg/api/annotation_key_constants.go]: Moved to azure_loadbalancer.go

**Comment by tomerf** [comment]
/retest
I don't see why the last test fails, ideas?

**Comment by tomerf** [comment]
All the tests pass, shall we continue?
@thockin @quinton-hoole @smarterclayton

**Comment by quinton-hoole** [comment]
Thanks @tomerf .  If this is going to be provider-specific I'd ideally prefer it to be implemented outside of the federation core (e.g. via some provider-specific controller that could watch for the annotations on ingress objects, and create the associated DNS records).  

Have you considered that approach?

**Comment by tomerf** [comment]
@quinton-hoole What do you mean?
Everything is done in the azure provider-specific code. This is handled once only on the Azure Frontend IP resource creation so no need for a controller to watch for the annotation.
I discussed the design proposal with @thockin and he prefers it is done that way.

**Comment by quinton-hoole** [comment]
@tomerf OK, thanks - let me take a closer look and think about it some more.  Will try to do this week.

**Comment by tomerf** [comment]
@quinton-hoole How's it going?

**Comment by quinton-hoole** [comment]
Sorry, I've been a bit swamped.  I haven't forgotten about you.

**Comment by quinton-hoole** [comment]
My apologies, I initially misread this as being related to federated services.  But it's for normal Kubernetes services.  I will defer to @kubernetes/sig-network-pr-reviews to decide on this one.  They should review, approve etc.

**Comment by cmluciano** [comment]
@kubernetes/sig-azure-misc

**Comment by tomerf** [comment]
@thockin ETA?

**Review by feiskyer — APPROVED** [review]
The change LGTM. ping @thockin for approval.

**Comment by tomerf** [comment]
/retest
Is there anything else needed to merge this?

**Pull request merged**
